app/src/behaviors/behavior_macro.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_macro
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+#include <zmk/behavior_queue.h>
+#include <zmk/keymap.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+enum behavior_macro_mode {
+    MACRO_MODE_TAP,
+    MACRO_MODE_PRESS,
+    MACRO_MODE_RELEASE,
+};
+
+struct behavior_macro_trigger_state {
+    uint32_t wait_ms;
+    uint32_t tap_ms;
+    enum behavior_macro_mode mode;
+    int8_t start_index;
+    int8_t end_index;
+};
+
+struct behavior_macro_state {
+    struct behavior_macro_trigger_state release_state;
+
+    int8_t last_press_binding;
```

Comment:
Maybe rename this to something like `press_end_index` for consistency with `release_state.end_index`? Also see comment above about types.

Alternatively, be even more consistent and have `struct behavior_macro_trigger_state press_state` which you set up once on init instead of in every `on_macro_binding_pressed()` call.

Reply:
Kept this, but renamed. Storing a whole struct to state that's always initialized to the same starter values seems wasteful, allocated on the stack when calling `queue_macro` instead.